![CC-BY-SA](https://mirrors.creativecommons.org/presskit/buttons/88x31/svg/by-sa.svg)


This notebook was created by [Bernardo Freitas Paulo da Costa](http://www.im.ufrj.br/bernardofpc),
and is licensed under Creative Commons BY-SA.

Antes de enviar este Teste, verifique que tudo está funcionando como esperado.
Por exemplo, **rode o código inteiro, do zero**.
Para isso, vá no menu, escolha _Kernel_, depois _Restart & Run All_.

Verifique, também, que você respondeu todas as questões:
* as questões de código têm `YOUR CODE HERE` (e você pode apagar o `raise NotImplemented` ao incluir sua resposta)
* as questões discursivas têm "YOUR ANSWER HERE".

---

# Teste 1: Bisseção

In [14]:
import numpy as np

## Questão 1: Número de iterações da bisseção, e de chamadas à função `f`

Generalize o algoritmo da bisseção para retornar, nesta ordem:
- a (aproximação da) raiz,
- o número de bisseções (divisões por 2 do intervalo) feitas,
- o número de vezes que você chamou a função `f`.

Use como critério de parada a tolerância `tol`,
ou seja, o algoritmo termina quando:
- seja possível garantir que o erro (absoluto) da resposta ("em $x$") seja menor do que `tol`.

In [2]:
def bissecao(f, a, b, tol=1e-8):
    z = (a + b) / 2
    niters = 1
    
    
    if (b - a) < tol:
        return z, niters
    
    
    if f(z)*f(a) < 0:
        tupla = bissecao(f, a, z, tol)
        
    else:
        tupla = bissecao(f, z, b, tol)
    
    raiz = tupla[0]
    niters += tupla[1]
    
    return raiz, niters, niters*2

### Testes simples

In [11]:
def p2(x): return x**2 - 2

In [4]:
bissecao(p2, 0, 2)

(1.414213564246893, 29, 58)

In [5]:
raiz, niters, ncalls = bissecao(p2, 0, 2)
assert abs(raiz - np.sqrt(2)) < 1e-8
assert 26 <= niters <= 30

In [6]:
raiz, niters, ncalls = bissecao(p2, 0, 2, tol=1e-3)
assert abs(raiz - np.sqrt(2)) < 1e-3
assert 10 <= niters <= 15

In [7]:
raiz, niters, ncalls = bissecao(p2, 0, 2, tol=1e-3)
assert abs(raiz - np.sqrt(2)) > 3e-4

In [8]:
# Testes escondidos aqui

### Testes mais legais

In [9]:
pimeios, niters, ncalls = bissecao(np.cos, 0, 2)
assert abs(pimeios - np.pi/2) <= 1e-8

In [10]:
menospimeios, niters, ncalls = bissecao(np.cos, -2, 0)
assert abs(menospimeios + np.pi/2) <= 1e-8

## Questão 2: várias iterações da bisseção

In [11]:
print(f'{"raiz":^9} {"niters"} {"ncalls"}, {"erro":^11} {"tol":^10}')
for p in range(10,20):
    raiz, niters, ncalls = bissecao(p2, 0, 2, tol=2**(-p))
    print(f"{raiz:.7f} {niters:6d} {ncalls:6d}, {raiz-np.sqrt(2): .8f} {2**(-p):.8f}")

  raiz    niters ncalls,    erro        tol    
1.4143066     13     26,  0.00009308 0.00097656
1.4141846     14     28, -0.00002899 0.00048828
1.4142456     15     30,  0.00003204 0.00024414
1.4142151     16     32,  0.00000153 0.00012207
1.4141998     17     34, -0.00001373 0.00006104
1.4142075     18     36, -0.00000610 0.00003052
1.4142113     19     38, -0.00000229 0.00001526
1.4142132     20     40, -0.00000038 0.00000763
1.4142141     21     42,  0.00000057 0.00000381
1.4142137     22     44,  0.00000010 0.00000191


O que você observa na tabela acima?  Explique.

Um acontecimento interessante que pode ser evidenciado pela tabela, é que nem sempre um número de iterações menor apresentará mais erro, como é evidenciado pelo erro da iteração 16 (0.00000153) e o erro da iteração 17 (0.00001373). E isso ocorre algumas vezes na lista.

É claro que ,normalmente , quanto mais iterações menor o erro.

E Não me parece trivial o porque disso.

## Questão 3: Outros critérios de parada

Modifique a sua bisseção para ter 2 critérios de parada:
- quando o tamanho do intervalo ficar menor do que `tol`
- quando o valor da função ficar menor do que `err`

In [12]:
def bissecao(f, a, b, tol=1e-8, err=1e-8):
    z = (a + b) / 2
    niters = 1
    
    if abs(f(z)) <= err:
        return z, niters, niters*2
    
    if (b - a) <= tol:
        return z, niters, niters*2

    if f(z)*f(a) < 0:
        tupla = bissecao(f, a, z, tol)

    else:
        tupla = bissecao(f, z, b, tol)

    raiz = tupla[0]
    niters += tupla[1]
    
    return raiz, niters, niters*2

Pare e pense:

uma solução iterativa pode ficar mais fácil de adaptar para vários critérios de parada
se ela for escrita com `while True`.

### Testes

In [16]:
raiz, niters, ncalls = bissecao(p2, 0, 2)
assert abs(p2(raiz) - 2) > 1e-8
assert abs(raiz - np.sqrt(2)) < 1e-8

In [17]:
assert bissecao(np.cos, 0, 2) != bissecao(np.cos, 0, 2, err=0)

AssertionError: 

In [18]:
assert bissecao(np.cos, 0, 2) == bissecao(np.cos, 0, 2, tol=0)

O que os dois testes acima indicam sobre a bisseção para `np.cos`?

YOUR ANSWER HERE

## Questão 4: Bisseção sem função

Escreva uma função que divide o intervalo ao meio, fica com o esquerdo,
e repete até o limite de precisão do computador.

Ela deve retornar o último intervalo não-trivial (como uma tupla), e o número de divisões.

In [26]:
def bisseção_esquerda(a,b):
    z = (a + b) / 2
    ndivs = 0
    
    if z == a:
        return (a, b), ndivs
    
    tupla = bisseção_esquerda(a, z)
    
    ndivs = tupla[1] + 1
    par = tupla[0]

    return par, ndivs

In [28]:
(an, bn), ndivs = bisseção_esquerda(1,2)
assert 2e-16 <= bn - an <= 4e-16
assert ndivs == 52
print(an, bn)

1 1.0000000000000002


In [27]:
(an, bn), ndivs = bisseção_esquerda(1,2)
assert 2e-16 <= bn - an <= 4e-16
assert ndivs == 52

In [21]:
(an, bn), ndivs = bisseção_esquerda(1,2)
assert 2e-16 <= bn - an <= 4e-16
assert ndivs == 52

In [ ]:
(an, bn), ndivs = bisseção_esquerda(bn,2)
assert 2e-16 <= bn - an <= 4e-16
assert ndivs == 51

Use esta função para calcular o próximo número de ponto flutuante.

In [22]:
def próximo_float(x):
    prox = x + 1
    tupla = bisseção_esquerda(x, prox)
    return tupla[0][1]

In [23]:
próximo_float(1)

1.0000000000000002

In [5]:
nextone = próximo_float(1)
assert nextone > 1
assert (1+nextone)/2 == 1

In [6]:
nextzero = próximo_float(0)
assert nextzero > 0
assert nextzero/2 == 0

In [ ]:
nextnext = próximo_float(próximo_float(1))
assert nextnext > 1
assert 1 < (nextnext+1)/2 < nextnext

In [7]:
maisum = próximo_float(-1.435)
assert maisum > -1.435
assert (maisum - 1.435)/2 == -1.435

## Questão 5: bisseção alternante

Escreva uma função que escolhe o lado esquerdo, e depois o lado direito, e repete, até acabar.

In [ ]:
def bisseção_alt(a,b):
    # YOUR CODE HERE
    raise NotImplementedError()

In [ ]:
(a,b), ndivs = bisseção_alt(1,2)
assert 1 < a < b < 2
assert ndivs == 52

In [ ]:
(a,b), ndivs = bisseção_alt(1,4)
assert 1 < a < b < 4
assert ndivs == 53
assert (a+b)/2 == b

In [ ]:
for (a0,b0) in [(1,2), (1,4), (2,4), (1/3,3/5)]:
    (a,b), ndivs = bisseção_alt(a0,b0)
    print(f"{a0:.5f} {b0:.5f} ---> {a:.17f} {b:.17f}")

O que esta função parece calcular?  Explique.

YOUR ANSWER HERE

In [ ]:
(a,b), ndivs = bisseção_alt(-1,2)
assert -1 < a < b < 2
assert (a+b)/2 == b
assert ndivs > 1000

Esta iteração levou mais de 1000 bisseções.  Explique o que aconteceu.

YOUR ANSWER HERE